staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/schema/cel/validation_test.go

```diff
@@ -0,0 +1,1012 @@
+					map[string]interface{}{"k": "a2", "v": "b2"},
+					map[string]interface{}{"k": "a3", "v": "b3", "v2": "z"},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"listMap": listMapType([]string{"k"}, objectTypePtr(map[string]schema.Structural{
+					"k":  stringType,
+					"v":  stringType,
+					"v2": stringType,
+				})),
+			}),
+			valid: []string{
+				"has(listMap[0].v)",
+				"listMap.all(m, m.k.startsWith('a'))",
+				"listMap.all(m, !has(m.v2) || m.v2 == 'z')",
+				"listMap.exists(m, m.k.endsWith('1'))",
+				"listMap.exists_one(m, m.k == 'a3')",
+				"!listMap.all(m, m.k.endsWith('1'))",
+				"!listMap.exists(m, m.v == 'x')",
+				"!listMap.exists_one(m, m.k.startsWith('a'))",
+				"size(listMap.filter(m, m.k == 'a1')) == 1",
+				"listMap.exists(m, m.k == 'a1' && m.v == 'b1')",
+				"listMap.map(m, m.v).exists(v, v == 'b1')",
+			},
+		},
+		{name: "list access",
+			obj: map[string]interface{}{
+				"array": []interface{}{1, 1, 2, 2, 3, 3, 4, 5},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"array": listType(&integerType),
+			}),
+			valid: []string{
+				"array.all(e, e > 0)",
+				"array.exists(e, e > 2)",
+				"array.exists_one(e, e > 4)",
+				"!array.all(e, e < 2)",
+				"!array.exists(e, e < 0)",
+				"!array.exists_one(e, e == 2)",
+				"array.all(e, e < 100)",
+				"size(array.filter(e, e%2 == 0)) == 3",
+				"array.map(e, e * 20).filter(e, e > 50).exists(e, e == 60)",
+				"size(array) == 8",
+			},
+		},
+		{name: "listSet access",
+			obj: map[string]interface{}{
+				"set": []interface{}{1, 2, 3, 4, 5},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"set": listType(&integerType),
+			}),
+			valid: []string{
+				"set.all(e, e > 0)",
+				"set.exists(e, e > 3)",
+				"set.exists_one(e, e == 3)",
+				"!set.all(e, e < 3)",
+				"!set.exists(e, e < 0)",
+				"!set.exists_one(e, e > 3)",
+				"set.all(e, e < 10)",
+				"size(set.filter(e, e%2 == 0)) == 2",
+				"set.map(e, e * 20).filter(e, e > 50).exists_one(e, e == 60)",
+				"size(set) == 5",
+			},
+		},
+
+		// Kubernetes special types
+		{name: "embedded object",
+			obj: map[string]interface{}{
+				"embedded": map[string]interface{}{
+					"apiVersion": "v1",
+					"kind":       "Pod",
+					"metadata": map[string]interface{}{
+						"name":         "foo",
+						"generateName": "pickItForMe",
+						"namespace":    "xyz",
+					},
+					"spec": map[string]interface{}{
+						"field1": "a",
+					},
+				},
+			},
+			schema: objectTypePtr(map[string]schema.Structural{
+				"embedded": {
+					Generic: schema.Generic{Type: "object"},
+					Extensions: schema.Extensions{
+						XEmbeddedResource: true,
+					},
+					Properties: map[string]schema.Structural{
+						"kind":       stringType,
+						"apiVersion": stringType,
+						"metadata": *objectTypePtr(map[string]schema.Structural{
+							"name":         stringType,
+							"generateName": stringType,
+						}),
+					},
```

Comment:
these fields are optional. Maybe another test-case with just `XEmbeddedResource` and even `embedded.kind` failing?

Reply:
maybe the test-case below has that (mixed with new properties).

Reply:
Agreed. This is fixed.

Reply:
I think this is still missing

Reply:
it's part of "embedded object with prune unknown", not really connected to "prune unknown". Might be worth to clarify.

Reply:
I've added some test cases and comments to make this more clear.